server/src/main/java/org/elasticsearch/common/lucene/uid/VersionsAndSeqNoResolver.java

```diff
@@ -193,4 +202,39 @@ public static long loadVersion(IndexReader reader, Term term) throws IOException
         final DocIdAndVersion docIdAndVersion = loadDocIdAndVersion(reader, term);
         return docIdAndVersion == null ? NOT_FOUND : docIdAndVersion.version;
     }
+
+    /**
+     * Looks up the primary term for a given seq_no in the provided directory reader. The caller must ensure that an operation with the
+     * given {@code seqNo} exists the provided {@code directoryReader}; otherwise this method will throw {@link IllegalStateException}.
+     */
+    public static long lookupPrimaryTerm(final DirectoryReader directoryReader, final long seqNo) throws IOException {
+        final DirectoryReader reader = Lucene.wrapAllDocsLive(directoryReader);
+        final IndexSearcher searcher = new IndexSearcher(reader);
+        searcher.setQueryCache(null);
+        final Query query = LongPoint.newExactQuery(SeqNoFieldMapper.NAME, seqNo);
+        final Weight weight = searcher.createWeight(query, ScoreMode.COMPLETE_NO_SCORES, 1.0f);
+        // iterate backwards since the existing operation is likely in the most recent segments.
+        for (int i = reader.leaves().size() - 1; i >= 0; i--) {
```

Comment:
is this optimization really relevant here? I wonder if we can't just do an ordinary search and then lookup the leaf reader based on the top hits?